test/sqllogictest/ldbc_bi.slt

```diff
@@ -0,0 +1,4022 @@
+    Return
+      Project (#0..=#2, #0, #4..=#6)
+        Filter (#2 <= 5) AND (#5 <= 5)
+          Map ((#1 + #4))
+            Join on=(#0 = #3) type=differential
+              ArrangeBy keys=[[#0]]
+                Reduce group_by=[#0] aggregates=[count(distinct #1), count(distinct #2)]
+                  Union
+                    Get l4
+                    Map (null)
+                      Union
+                        Negate
+                          Distinct group_by=[#0, #1]
+                            Project (#0, #1)
+                              Get l4
+                        Get l2
+              ArrangeBy keys=[[#0]]
+                Reduce group_by=[#0] aggregates=[count(distinct #1), count(distinct #2)]
+                  Union
+                    Get l6
+                    Map (null)
+                      Union
+                        Negate
+                          Distinct group_by=[#0, #1]
+                            Project (#0, #1)
+                              Get l6
+                        Get l5
+    With
+      cte l6 =
+        Project (#0, #1, #3)
+          Join on=(#0 = #2 AND #3 = #4) type=differential
+            ArrangeBy keys=[[#0]]
+              Get l5
+            Get l3
+            ArrangeBy keys=[[#0]]
+              Distinct group_by=[#0]
+                Project (#0)
+                  Get l5
+      cte l5 =
+        Project (#0, #2)
+          Join on=(#0 = #1 AND #2 = #3) type=differential
+            Get l0
+            ArrangeBy keys=[[#0]]
+              Distinct group_by=[#1, #0]
+                Project (#1, #9)
+                  Filter (2012-12-14 00:00:00 = date_to_timestamp(timestamp_with_time_zone_to_date(#0)))
+                    Get materialize.public.message
+            ArrangeBy keys=[[#0]]
+              Distinct group_by=[#0]
+                Project (#0)
+                  Filter (#3 = "Thailand_Noriega") AND (#1) IS NOT NULL
+                    Get l1
+      cte l4 =
+        Project (#0, #1, #3)
+          Join on=(#0 = #2 AND #3 = #4) type=differential
+            ArrangeBy keys=[[#0]]
+              Get l2
+            Get l3
+            ArrangeBy keys=[[#0]]
+              Distinct group_by=[#0]
+                Project (#0)
+                  Get l2
+      cte l3 =
+        ArrangeBy keys=[[#1]]
+          Project (#1, #2)
+            Get materialize.public.person_knows_person
+      cte l2 =
+        Project (#0, #2)
+          Join on=(#0 = #1 AND #2 = #3) type=differential
+            Get l0
+            ArrangeBy keys=[[#0]]
+              Distinct group_by=[#1, #0]
+                Project (#1, #9)
+                  Filter (2012-10-07 00:00:00 = date_to_timestamp(timestamp_with_time_zone_to_date(#0)))
+                    Get materialize.public.message
+            ArrangeBy keys=[[#0]]
+              Distinct group_by=[#0]
+                Project (#0)
+                  Filter (#3 = "Diosdado_Macapagal") AND (#1) IS NOT NULL
```

Comment:
Stuck again. And in this situation it's clearly that we did `RelationCSE` before `PredicatePushdown`. Or alternatively that `PredicatePushdown` doesn't undo the CSE, which it could do with cost-based optimization after seeing that the 2 new things would be much cheaper in total than the one original thing.

Reply:
Nice! Added a note about this to the design doc.